stellargraph/core/element_data.py

```diff
@@ -35,7 +35,9 @@ class ExternalIdIndex:
 
     def __init__(self, ids):
         self._index = pd.Index(ids)
-        self._dtype = np.min_scalar_type(len(self._index))
+        self._dtype = np.min_scalar_type(
+            len(self._index) + 1
+        )  # reserve 2 ^ (n-bits) - 1 for sentinel
```

Comment:
This was already reserving the highest element and the `test_external_id_index_to_iloc` test is verifying it (although not explicitly). This because the `np.min_scalar_type` argument is inclusive, and we never have an iloc equal to `len(self._index)` (since `len(self._index) - 1` is the largest).

```python
import numpy as np
np.min_scalar_type(255) # dtype('uint8')
np.min_scalar_type(256) # dtype('uint16')
```

That is, if we have 256 elements, we could theoretically fit them all into a `uint8` but we don't do that because what we pass here.

However, it'd definitely be nice to note this:

```suggestion
        # reserve 2 ^ (n-bits) - 1 for sentinel
        self._dtype = np.min_scalar_type(len(self._index)) 
```

(Note that I've put the comment beforehand, so that `black` doesn't format it unnecessarily complicated-ly.)